hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MapIndexScanP.java

```diff
@@ -0,0 +1,411 @@
+                    extremeIndex = i;
+                    extreme = s.currentRow;
+                }
+            }
+
+            if (extremeIndex < 0) {
+                assert splits.isEmpty();
+                return true;
+            }
+
+            pendingItem = extreme;
+            splits.get(extremeIndex).remove();
+        }
+    }
+
+    private boolean runHashIndex() {
+        for (; ; ) {
+            for (int i = 0; i < splits.size(); ++i) {
+                Split s = splits.get(i);
+                try {
+                    s.peek();
+                } catch (MissingPartitionException e) {
+                    splits.addAll(splitOnMigration(s));
+                    splits.remove(i--);
+                    continue;
+                }
+                if (s.currentRow == null) {
+                    if (s.done()) {
+                        // No more items to read, remove finished split.
+                        splits.remove(i--);
+                        if (splits.isEmpty()) {
+                            return true;
+                        }
+                    }
+                } else {
+                    if (tryEmit(s.currentRow)) {
+                        s.remove();
+                    } else {
+                        return false;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Perform splitting of a {@link Split} after receiving {@link MissingPartitionException}.
+     * <p>
+     * Method gets current partition table and proceeds with following procedure:
+     * <p>
+     * It splits the partitions assigned to the split according to the new
+     * partition owner into disjoint sets, one for each owner.
+     *
+     * @param split the split to split
+     * @return collection of new split units
+     */
+    private List<Split> splitOnMigration(Split split) {
+        IndexIterationPointer[] lastPointers = split.pointers;
+        InternalPartitionService partitionService = getNodeEngine(hazelcastInstance).getPartitionService();
+        PrimitiveIterator.OfInt partitionIterator = split.partitions.intIterator();
+        Map<Address, Split> newSplits = new HashMap<>();
+        while (partitionIterator.hasNext()) {
+            int partitionId = partitionIterator.nextInt();
+            Address owner = partitionService.getPartition(partitionId).getOwnerOrNull();
+            // TODO: does it possible that owner is NULL?
+            newSplits.computeIfAbsent(owner, x -> new Split(
+                    new PartitionIdSet(partitionService.getPartitionCount()), owner, lastPointers)
+            ).partitions.add(partitionId);
+        }
+        return new ArrayList<>(newSplits.values());
+    }
+
+    private Object[] projectAndFilter(@Nonnull QueryableEntry<?, ?> entry) {
+        /** Was added to prevent NPE in any {@link QueryableEntry} implementor. **/
+        entry.setSerializationService(evalContext.getSerializationService());
```

Comment:
There could be a situation where processor receives a `QueryableEntry` object with null `serializationService`.
For the moment this is the best way to avoid NPE and save correct semantics.

Reply:
How and why is it possible?

Reply:
How did I found such an issue : I detected `CacheQueryEntry` object will `null` `serializationService` during migration test launch.
How does it possible : for example, somewhere it was created by empty ctor, and used somewhere in `MapFetchIndexOperation` or or somewhere deeper (in Map internals or whatever).

Please, note that this is very sporadic behavior and I found it in 1-2 from 10 runs of stress test.

Reply:
It feels like the problem lies somewhere else (or at least we don't understand what's the real culprit here). I think we should first figure out the root cause and not trying to hack around it.

Reply:
Successfully fixed.